File: frontend/src/lib/wireguardStatus.ts
import { parseTimestamp } from '@/lib/datetime'
import type { WireGuardPeer } from '@/types'

/** WireGuard marks peers stale after ~3 minutes without handshake. */
const WG_HANDSHAKE_ONLINE_SECONDS = 180

export function isWireGuardOnline(peer: WireGuardPeer) {
  if (!peer.latest_handshake) return false
  const parsed = parseTimestamp(peer.latest_handshake)
  if (!parsed) return false
  const handshakeMs = parsed.getTime()
  if (Number.isNaN(handshakeMs)) return false
  const ageSeconds = (Date.now() - handshakeMs) / 1000
  return ageSeconds >= 0 && ageSeconds <= WG_HANDSHAKE_ONLINE_SECONDS
}

File: backend/app/services/wireguard_status.py
"""WireGuard peer online detection helpers."""

from __future__ import annotations

from datetime import datetime, timezone

from app.schemas import WireGuardPeer

# WireGuard considers a peer stale after ~3 minutes without traffic/handshake.
WG_HANDSHAKE_ONLINE_SECONDS = 180


def wireguard_peer_is_online(
    peer: WireGuardPeer,
    *,
    max_age_seconds: int = WG_HANDSHAKE_ONLINE_SECONDS,
) -> bool:
    handshake = peer.latest_handshake
    if not handshake:
        return False
    try:
        parsed = datetime.fromisoformat(str(handshake))
        if parsed.tzinfo is not None:
            parsed = parsed.astimezone(timezone.utc).replace(tzinfo=None)
    except (TypeError, ValueError):
        return False
    age_seconds = (datetime.utcnow() - parsed).total_seconds()
    return 0 <= age_seconds <= max_age_seconds
